Rewrite the Quadlet documentation

jankaluza opened the pull request:
This commit does the following:
  
  - Splits the podman-systemd.unit.5.md into multiple files - one for each quadlet file type.
  - Adds the podman-quadlet-basic-usage.7.md for quadlet examples.
  - Majority of the text in the new files is copied from the podman-systemd.unit.5.md
  - Adds support for very simple condditional in the markdown_preprocess.
  - Uses new logic in markdown_preprocess in options/*.md to use a single .md file for both
    podman subcommands man-pages and quadlet man-pages. This deduplicates the Quadlet man-pages a lot.
  - Adds new `@@option quadlet:source.md`` preprocess command to import such .md files from options directory.
  

<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
None
```

TomSweeneyRedHat commented on docs/source/markdown/options/README.md:
```diff
@@ -25,6 +27,41 @@ file, this script creates a `.md` file that can then be read by
+Conditionals
+============
+
+Some options are used as both Podman command line option and Quadlet
```
```suggestion
Some options are used as both Podman command line options and Quadlet
```

TomSweeneyRedHat commented on docs/source/markdown/options/dns.md:
```diff
@@ -1,15 +1,19 @@
 configuration passed to the container. Typically this is necessary when the
 host DNS configuration is invalid for the container (e.g., **127.0.0.1**). When this
-is the case the **--dns** flag is necessary for every run.
+is the case the << '**DNS=.**' if is_quadlet else '**--dns**' >> flag is necessary for every run.
```
It might be md ignorance on my part, but I'm not sure what this line does.

TomSweeneyRedHat replied:
looking at examples below, shouldn't the `if` be  the first thing here?

TomSweeneyRedHat commented on docs/source/markdown/options/exec.md:
```diff
@@ -0,0 +1,18 @@
+more arguments after a `podman run <image> <arguments>` invocation.
+
+The format is the same as for [systemd command lines](https://www.freedesktop.org/software/systemd/man/systemd.service.html#Command%20lines),
+However, unlike the usage scenario for similarly-named systemd `ExecStart=` verb
```
```suggestion
However, unlike the usage scenario for similarly-named systemd `ExecStart=` verbs
```

jankaluza replied:
Shouldn't it be a singular? There's just single `ExecStart` verb. If it should really be a plural, I will also change the `operates` to `operate` in the next line, ok?

mheon replied:
Agree singular is more appropriate

TomSweeneyRedHat commented on docs/source/markdown/options/file.md:
```diff
@@ -1,16 +1,31 @@
 last specified file.
 
+<< if is_quadlet >>
+Note that for a given relative path to a Containerfile, or when using a `http(s)://` URL, you also must set
```
```suggestion
Note that for a given relative path to a Containerfile, or when using a `http(s)://` URL, also set
```
We try to avoid pronouns like you in man pages.

TomSweeneyRedHat commented on docs/source/markdown/options/internal.md:
```diff
@@ -0,0 +1,25 @@
+When using the `macvlan` or `ipvlan` driver with this option no default route will be added to the container.
+Because it bypasses the host network stack no additional restrictions can be set by podman and if a
+privileged container is run it can set a default route themselves. If this is a concern then the
+container connections should be blocked on your actual network gateway.
```
```suggestion
container connections should be blocked on your actual network gateway.
```
```suggestion
container connections should be blocked on the actual network gateway.
```

TomSweeneyRedHat commented on docs/source/markdown/options/sysctl.md:
```diff
@@ -18,6 +22,7 @@ For the IPC namespace, the following sysctls are allowed:
 - kernel.shm_rmid_forced
 - Sysctls beginning with fs.mqueue.\*
 
+
```
is this extra line needed?

Luap99 commented on docs/source/markdown/podman-build.unit.5.md.in:
```diff
@@ -0,0 +1,176 @@
+- [podman-quadlet(7)](podman-quadlet.7.md)
+- [systemd.unit(5)](https://www.freedesktop.org/software/systemd/man/systemd.unit.html)
+- [podman-build(1)](https://docs.podman.io/en/latest/markdown/podman-build.1.html)
+- [podman-quadlet(7)]
```
not a proper link and duplicated anyway

Luap99 commented on docs/source/markdown/podman-container.unit.5.md.in:
```diff
@@ -0,0 +1,502 @@
+
+[systemd.unit(5)](https://www.freedesktop.org/software/systemd/man/systemd.unit.html),
+[systemd.service(5)](https://www.freedesktop.org/software/systemd/man/systemd.service.html),
+[systemd-analyze(1)](https://www.freedesktop.org/software/systemd/man/latest/systemd-analyze.html),
```
why is systemd-analyze linked here? the main page described how to debug with it not this file

Luap99 commented on docs/source/markdown/podman-container.unit.5.md.in:
```diff
@@ -0,0 +1,502 @@
+[systemd.service(5)](https://www.freedesktop.org/software/systemd/man/systemd.service.html),
+[systemd-analyze(1)](https://www.freedesktop.org/software/systemd/man/latest/systemd-analyze.html),
+[podman-run(1)](https://docs.podman.io/en/latest/markdown/podman-run.1.html),
+[podman-quadlet(7)]
```
not a proper link

Luap99 commented on docs/source/markdown/podman-image.unit.5.md.in:
```diff
@@ -0,0 +1,132 @@
+
+[systemd.unit(5)](https://www.freedesktop.org/software/systemd/man/systemd.unit.html),
+[podman-image-pull(1)](https://docs.podman.io/en/latest/markdown/podman-image-pull.1.html),
+[podman-quadlet(7)]
```
not a proper link

Luap99 commented on docs/source/markdown/podman-network.unit.5.md.in:
```diff
@@ -0,0 +1,127 @@
+
+[systemd.unit(5)](https://www.freedesktop.org/software/systemd/man/systemd.unit.html),
+[podman-network-create(1)](https://docs.podman.io/en/latest/markdown/podman-network-create.1.html),
+[podman-quadlet(7)]
```
not a proper link

Luap99 commented on docs/source/markdown/podman-pod.unit.5.md.in:
```diff
@@ -0,0 +1,146 @@
+
+[systemd.unit(5)](https://www.freedesktop.org/software/systemd/man/systemd.unit.html),
+[podman-kube-play(1)](https://docs.podman.io/en/latest/markdown/podman-kube-play.1.html),
+[podman-quadlet(7)]
```
again here

Luap99 reviewed (dismissed):
Again, like I commented before on the other PR and I had to fix this up in https://github.com/containers/podman/pull/27002

Never delete podman-systemd.unit.5 please. It should contain the content of podman-quadlet.7 and there should be never a podman-quadlet.7 IMO. It is confusing since there is podman-quadlet.1 already.

I have not double check the actual content

jankaluza committed:
Rewrite the Quadlet documentation

This commit does the following:

- Splits the podman-systemd.unit.5.md into multiple files - one for each quadlet file type.
- Adds the podman-quadlet-basic-usage.7.md for quadlet examples.
- Majority of the text in the new files is copied from the podman-systemd.unit.5.md
- Adds support for very simple condditional in the markdown_preprocess.
- Uses new logic in markdown_preprocess in options/*.md to use a single .md file for both
  podman subcommands man-pages and quadlet man-pages. This deduplicates the Quadlet man-pages a lot.
- Adds new `@@option quadlet:source.md`` preprocess command to import such .md files from options directory.

Signed-off-by: Jan Kaluza <jkaluza@redhat.com>

mheon commented:
Can this get a rebase? I would love to get this landed soon, so it's not forgotten about.

lsm5 commented:
@jankaluza needs rebase.

Honny1 commented on docs/source/markdown/options/restart.md:
```diff
@@ -19,4 +19,4 @@ Podman provides a systemd unit file, podman-restart.service, which restarts cont
 When running containers in systemd services, use the restart functionality provided by systemd.
 In other words, do not use this option in a container unit, instead set the `Restart=` systemd directive in the `[Service]` section.
-See **podman-systemd.unit**(5) and **systemd.service**(5).
+See **podman-systemd**(5) and **systemd.service**(5).
```
Is this the correct change?

Honny1 commented on docs/source/markdown/options/dns-search.container.md:
```diff
@@ -1,8 +1,13 @@
-Set custom DNS search domains. Invalid if using **--dns-search** with **--network** that is set to **none** or **container:**_id_.
-Use **--dns-search=.** to remove the search domain.
+Set custom DNS search domains. Invalid if using << '**DNSSearch=**' if is_quadlet else '**--dns-search**' >>
+with with << '**Network=**' if is_quadlet else '**--network**' >> that is set to **none** or **container:**_id_.
```
```suggestion
with << '**Network=**' if is_quadlet else '**--network**' >> that is set to **none** or **container:**_id_.
```

Honny1 commented on docs/source/markdown/options/entrypoint.md:
```diff
@@ -12,7 +16,7 @@ because it specifies what executable to run when the container starts, but it is
 container runs as if it were that binary, complete with default options. More options can be
 passed in via the COMMAND. But, if a user wants to run
-something else inside the container, the **--entrypoint** option allows a new
+something else inside the container, the << '**Entrypoint=**' if is_quadlet else '**--entrypoint=.**' >>option allows a new
```
```suggestion
something else inside the container, the << '**Entrypoint=**' if is_quadlet else '**--entrypoint=.**' >> option allows a new
```

Honny1 reviewed (commented):
I found some small nits. Expect CI complaints.

Honny1 reviewed (approved):
LGTM

PTAL @Luap99 @mheon @containers/podman-maintainers

inknos commented on docs/source/markdown/options/entrypoint.md:
```diff
@@ -12,7 +16,7 @@ because it specifies what executable to run when the container starts, but it is
 container runs as if it were that binary, complete with default options. More options can be
 passed in via the COMMAND. But, if a user wants to run
-something else inside the container, the **--entrypoint** option allows a new
+something else inside the container, the << '**Entrypoint=**' if is_quadlet else '**--entrypoint=.**' >> option allows a new
```
```suggestion
something else inside the container, the << '**Entrypoint=**' if is_quadlet else '**--entrypoint=**' >> option allows a new
```

inknos commented on docs/source/markdown/options/read-only-tmpfs.md:
```diff
@@ -1,8 +1,12 @@
 ####> If file is edited, make sure the changes
 ####> are applicable to all of those.
+<< if is_quadlet >>
+### `ReadOnlyTmpfs=`
```
there's not a value hint here. should it?

inknos commented on docs/source/markdown/options/configmap.md:
```diff
@@ -0,0 +1,18 @@
+#### **--configmap**=*path*
+<< endif >>
+
+Use Kubernetes configmap YAML at path to provide a source for environment variable values within the containers of the pod.  (This option is not available with the remote Podman client)
```
```suggestion
Use Kubernetes configmap YAML as path to provide a source for environment variable values within the containers of the pod.  (This option is not available with the remote Podman client)
```

TomSweeneyRedHat replied:
```suggestion
Use the Kubernetes ConfigMap YAML as the path to provide environment variable values within the containers of the pod.  (This option is not available with the remote Podman client)
```

inknos commented on docs/source/markdown/options/driver.md:
```diff
@@ -0,0 +1,25 @@
+<< endif >>
+
+Driver to manage the network. Currently `bridge`, `macvlan` and `ipvlan` are supported. Defaults to `bridge`.
+When running rootless, the `macvlan` and `ipvlan` driver have no access to the host network interfaces because rootless networking requires a separate network namespace.
```
```suggestion
When running rootless, the `macvlan` and `ipvlan` drivers have no access to the host network interfaces because rootless networking requires a separate network namespace.
```

eriksjolund commented on docs/source/markdown/podman-kube.unit.5.md.in:
```diff
@@ -0,0 +1,131 @@
+
+### `SetWorkingDirectory=yaml`
+
+Set the `WorkingDirectory` directive in the `Service` section of the systemd service unit file.
```
`Service` section 

->

`[Service]` section

TomSweeneyRedHat commented on docs/source/markdown/options/configmap.md:
```diff
@@ -0,0 +1,18 @@
+<< if is_quadlet >>
+The value may contain only one path but it may be absolute or relative to the location of the unit file.
+<< else >>
+Note: The **--configmap** option can be used multiple times or a comma-separated list of paths can be used to pass multiple Kubernetes configmap YAMLs.
```
```suggestion
Note: The **--configmap** option can be used multiple times, or a comma-separated list of paths can be used to pass multiple Kubernetes ConfigMap YAMLs.
```
It probably should be "Kubernetes ConfigMap" throughout.

TomSweeneyRedHat commented on docs/source/markdown/options/internal.md:
```diff
@@ -0,0 +1,25 @@
+Restrict external access of this network when using a `bridge` network. Note when using the CNI backend
+DNS will be automatically disabled, see **--disable-dns**.
+
+When using the `macvlan` or `ipvlan` driver with this option no default route will be added to the container.
```
```suggestion
When using the `macvlan` or `ipvlan` driver with this option, no default route will be added to the container.
```

TomSweeneyRedHat commented on docs/source/markdown/options/internal.md:
```diff
@@ -0,0 +1,25 @@
+DNS will be automatically disabled, see **--disable-dns**.
+
+When using the `macvlan` or `ipvlan` driver with this option no default route will be added to the container.
+Because it bypasses the host network stack no additional restrictions can be set by podman and if a
```
```suggestion
Because it bypasses the host network stack, no additional restrictions can be set by Podman, and if a
```

TomSweeneyRedHat commented on docs/source/markdown/options/internal.md:
```diff
@@ -0,0 +1,25 @@
+
+When using the `macvlan` or `ipvlan` driver with this option no default route will be added to the container.
+Because it bypasses the host network stack no additional restrictions can be set by podman and if a
+privileged container is run the container can set a default route itself. If this is a concern then the
```
```suggestion
privileged container is run, the container can set a default route itself. If this is a concern, then the
```

TomSweeneyRedHat commented on docs/source/markdown/options/ipam-driver.md:
```diff
@@ -0,0 +1,22 @@
+#### **--ipam-driver**=*driver*
+<< endif >>
+
+Set the ipam driver (IP Address Management Driver) for the network. When unset podman chooses an
```
```suggestion
Set the ipam driver (IP Address Management Driver) for the network. When unset, Podman chooses an
```

TomSweeneyRedHat commented on docs/source/markdown/options/ipam-driver.md:
```diff
@@ -0,0 +1,22 @@
+
+Valid values are:
+
+ - `dhcp`: IP addresses are assigned from a dhcp server on the network. When using the netavark backend
```
```suggestion
 - `dhcp`: IP addresses are assigned from a DHCP server on the network. When using the netavark backend
```

When referencing DHCP, it should be in caps, other than the value here.

TomSweeneyRedHat commented on docs/source/markdown/options/ipv6.md:
```diff
@@ -0,0 +1,11 @@
+#### **--ipv6**
+<< endif >>
+
+Enable IPv6 (Dual Stack) networking. If no subnets are given, it allocates an ipv4 and an ipv6 subnet.
```
```suggestion
Enable IPv6 (Dual Stack) networking. If no subnets are given, it allocates an IPv4 and an IPv6 subnet.
```
Please make similar changes throughout.

TomSweeneyRedHat commented on docs/source/markdown/options/opt.md:
```diff
@@ -0,0 +1,40 @@
+- `isolate`: This option isolates networks by blocking traffic between those that have this option enabled.
+- `com.docker.network.bridge.name`: This option assigns the given name to the created Linux Bridge
+- `com.docker.network.driver.mtu`: Sets the Maximum Transmission Unit (MTU) and takes an integer value.
+- `vrf`: This option assigns a VRF to the bridge interface. It accepts the name of the VRF and defaults to none. Can only be used with the Netavark network backend.
```
```suggestion
- `vrf`: This option assigns a Virtual routing and forwarding (VRF) routing table to the bridge interface. It accepts the VRF name and defaults to none. Can only be used with the Netavark network backend.
```
I think?

TomSweeneyRedHat commented on docs/source/markdown/options/stop-timeout.md:
```diff
@@ -1,8 +1,16 @@
 Remote connections use local containers.conf for defaults.
+
+<< if is_quadlet >>
+Note, this value should be lower than the actual systemd unit timeout to make sure the `podman rm` command is not killed by systemd.
```
```suggestion
Note: this value should be lower than the actual systemd unit timeout to ensure the `podman rm` command is not killed by systemd.
```

TomSweeneyRedHat replied:
IDK, but maybe bold the word "Note".

TomSweeneyRedHat commented on docs/source/markdown/options/subgidname.md:
```diff
@@ -1,11 +1,14 @@
-Together with **--gidmap** it acts as if **--gidmap** was passed with all
-mappings of the _name_ group in _/etc/subgid_.
+Run the container in a new user namespace using the map with _name_ in the _/etc/subgid_ file.
+If running rootless, the user needs to have the right to use the mapping. See **subgid**(5).
```
```suggestion
When running rootless, the user needs the necessary permissions to use the mapping. See **subgid**(5).
```

TomSweeneyRedHat commented on docs/source/markdown/options/subnet.md:
```diff
@@ -0,0 +1,15 @@
+<< if not is_quadlet >>
+The argument order of the **--subnet**, **--gateway** and **--ip-range** options must match.
+<< endif >>
+This is useful to set a static ipv4 and ipv6 subnet.
```
```suggestion
This is useful to set a static IPv4 and IPv6 subnet.
```

TomSweeneyRedHat commented on docs/source/markdown/options/subuidname.md:
```diff
@@ -1,11 +1,14 @@
-Together with **--uidmap** it acts as if **--uidmap** was passed with all
-mappings of the _name_ user in _/etc/subuid_.
+Run the container in a new user namespace using the map with _name_ in the _/etc/subuid_ file.
+If running rootless, the user needs to have the right to use the mapping. See **subuid**(5).
```
```suggestion
When running rootless, the user needs the necessary permissions to use the mapping. See **subuid**(5).
```

TomSweeneyRedHat commented on docs/source/markdown/.gitignore:
```diff
@@ -72,3 +73,11 @@ podman-unpause.1.md
+podman-kube-down.1.md
+podman-network.unit.5.md
+podman-pod.unit.5.md
+podman-volume.unit.5.md
```
To keep my OCD from freaking out, please alphabetize these

TomSweeneyRedHat commented on docs/source/markdown/podman-image.unit.5.md.in:
```diff
@@ -0,0 +1,132 @@
+
+### `ImageTag=`
+
+The actual FQIN of the referenced `Image`.
```
```suggestion
The actual Fully Qualified Image Name (FQIN) of the referenced `Image`.
```
I don't see this spelled out elsewhere in this doc, so it's probably good to spell it out at the first reference.

TomSweeneyRedHat commented on docs/source/markdown/podman-network-create.1.md.in:
```diff
@@ -0,0 +1,114 @@
+#### **--ignore**
+
+Ignore the create request if a network with the same name already exists instead of failing.
+Note, trying to create a network with an existing name and different parameters does not change the configuration of the existing one.
```
```suggestion
Note: trying to create a network with an existing name and different parameters does not change the configuration of the existing one.
```

If you bold "Note" above, bold here too.

TomSweeneyRedHat commented on docs/source/markdown/podman-systemd.unit.5.md:
```diff
@@ -8,6 +8,14 @@ podman\-systemd.unit - systemd units using Podman Quadlet
+- **`.network`** — Creates a Podman network for containers and pods. See [podman-network.unit(5)](podman-network.unit.5.md).
+- **`.image`** — Pulls and caches a container image. See [podman-image.unit(5)](podman-image.unit.5.md).
+- **`.build`** — Builds a container image from a Containerfile. See [podman-build.unit(5)](podman-build.unit.5.md).
+- **`.kube`** — Deploys containers from Kubernetes YAML using [podman-kube.unit(5)](podman-kube.unit.5.md).
```
OCD check, can these be alphabetized?

TomSweeneyRedHat commented on hack/xref-quadlet-docs:
```diff
@@ -19,7 +19,16 @@ our $VERSION = '0.1';
+    "docs/source/markdown/podman-pod.unit.5.md",
+    "docs/source/markdown/podman-volume.unit.5.md",
+    "docs/source/markdown/podman-image.unit.5.md",
+);
```
Another OCD alphabetization request

Honny1 reviewed (approved):
LGTM
